### PR TITLE
feature/vcm-7209: Analytics Ordering

### PR DESCRIPTION
**Pull request by ursaspace**
# VCM-7209: Analytics Ordering

## Whats Changed
- New `Analytic` Tool in Toolbox (w/ XML file for tool metadata)
- Moved custom arcpy parameters to `parameters` directory
- Moved all source code files into `src` directory (less confusing to non-technical folks who need to install this)
- Updated root README, moved technical details to separate readme in `src`

## Notes
- Ran prettier on all files and rearranged the project structure (so more changes than usual)
- Trevor wrote the README, so added some more details on the analytics ordering following his pattern
- Ordering works with all internal accounts and external L2 accounts that are setup for invoicing. L1 accounts or external accounts that are not setup for invoicing will fail to submit an order.
- The UI controls in the toolbox are pretty limited. Followed the same behaviors that were approved for the tasking tool. Not many other improvements we can make to the UX.
- You will need access to the ArcGIS Instance in AWS (reach out to IT for access, say its for the toolbox). You'll also need an ArcGIS login with a license to ArcGIS Pro (IT can set you up with that as well)

## To Test
- Run tests (see readme)
- Submit an analytic order! See confirmation email with order details

Screenshot 2023-05-01 at 10 38 08 AM

### CODE REVIEW

**Comment by samanthaweeks** [comment]
We should provide the disclaimer that the radius can only be between 1 and 5 ahead of time because how else would I have known I couldn't pick 60000000 

Screenshot 2023-05-02 at 9 25 55 AM

**Comment by samanthaweeks** [comment]
I know the start and end date have to be 2 days apart, but if I changed the start date to be later than the end date (rather than changing the end date) should we display the error on the start field since that's the one I edited? It can be the same exact message, because it's the same error, but I think we should error on whichever field the user actually touched

Screenshot 2023-05-02 at 9 26 25 AM

**Comment by samanthaweeks** [comment]
When I empty either the start or end date, all the fields error, but that doesn't happen with any of the other fields, so that feels a little weird to me. I think it's fair to mark the field I messed with as invalid, but not the other ones (the field I actually touched it the only one that isn't showing errored)

Screenshot 2023-05-02 at 9 27 18 AM


here's the error message I get when i hover 
Screenshot 2023-05-02 at 9 38 29 AM

**Comment by samanthaweeks** [comment]
These should be listed in the same order, they don't quite match

Screenshot 2023-05-02 at 9 28 01 AM

**Comment by samanthaweeks** [comment]
The change detection options list says it's optional, but when I submit a job without selecting anything from there it errors, so it's... not optional?? 

here's the error message I get

Screenshot 2023-05-02 at 9 31 38 AM

**Comment by samanthaweeks** [comment]
I think this is a backend change since we're not handling sending the emails, but the email thanks me for my imagery order, but I'm ordering analytics - we should probably display the actual thing the user ordered otherwise they might get confused and cancel their order if they think it went through wrong

File: README.md [hunk 1]
```diff
@@ -1,6 +1,7 @@
 # Ursa's ArcGIS Python Toolbox
 
 ## Ursa Overview
+
 Ursa Space maintains robust partnerships with existing and upcoming commercial data providers, and regularly ingests and processes a variety of SAR and non-SAR (i.e. optical, infrared, multi-spectral, and hyperspectral) data. This collection of data partners is referred to as Ursa’s Virtual Constellation.
 
 Ursa Space performs this processing through the development and deployment of the Ursa Platform. The platform contains the data and software services required to ingest data (SAR imagery, optical imagery, external data sources for algorithmic fusion, etc.), run analytics and algorithms on these data to produce results, and provide customer access to related data feeds via API.
```
Comment (samanthaweeks) [inline README.md:7]: I know you didn't write this, so feel free to say you want to keep it as is, but

```suggestion
Ursa Space performs this processing through the development and deployment of the Ursa Platform. The platform contains the data and software services required to ingest data, run analytics and algorithms on these data to produce results, and provide customer access to related data feeds via API.
```

we already list what that data is above so it feels redundant to me

File: README.md [hunk 2]
```diff
@@ -15,49 +16,63 @@ A python toolbox for working with the Ursa Platform APIs in ArcGIS Pro.
...
+4. Open the added toolbox. Listed are the available tools.
+   <img width="323" alt="Screenshot 2023-01-30 at 15 16 21" src="https://user-images.githubusercontent.com/123033437/215618493-349f235f-2d0a-47cb-877e-e78138ab46e4.png">
 
 ## Usage
+
 Tools can be opened by double-clicking the item in the toolbox.
 
 ### Login
+
 - User must provide Ursa Platform credentials using the "Login" tool in order to use the Ursa Platform APIs.
```
Comment (samanthaweeks) [inline README.md:50]: I would specify what happens if you try to use one of the order tools when you're not logged in - that way if anyone doesn't realize they didn't log in and gets the error we mention here, they'll easily be able to tell "oh, it's just that I didn't log in" and it'll be more clear that it's not some other sort of error

File: README.md [hunk 3]
```diff
@@ -15,49 +16,63 @@ A python toolbox for working with the Ursa Platform APIs in ArcGIS Pro.
...
 - End Date cannot precede start date
 - Determine Imaging Mode
 - Determine Preferred Vendor(s), _future iterations will include additional vendors_.
 
+### Analytics
+
+- Area of interest only supports point geometries
+- End Date cannot precede start data
+- Determine Analytic Options
+- If Change Detection Analytic selected, determine specific Change Detection Analytic options
```
Comment (samanthaweeks) [inline README.md:66]:
```suggestion
- If Change Detection Analytic selected, determine specific Change Detection type from the Change Detection Analytic options
```

**Comment by ursaspace** [comment]
> We should provide the disclaimer that the radius can only be between 1 and 5 ahead of time because how else would I have known I couldn't pick 60000000
> 
> Screenshot 2023-05-02 at 9 25 55 AM

Ok added to the parameter description

**Comment by ursaspace** [comment]
> I know the start and end date have to be 2 days apart, but if I changed the start date to be later than the end date (rather than changing the end date) should we display the error on the start field since that's the one I edited? It can be the same exact message, because it's the same error, but I think we should error on whichever field the user actually touched

Two things:
1) I think this is ok since (in this example) the user specifically wants that start date. The end date here is what is now invalid, not the start the date.

2) There isn't a good way to know which parameter actually changed. Whenever any one of the parameters change, those validators are ran. So in the validators, you can't know which value actually changed, just what the latest value from that parameter is, and you can't access previous values. I think thats just a limitation of these arcpy parameters. So in this example, we won't know whether the start date or the end date was what has changed.

**Comment by ursaspace** [comment]
> These should be listed in the same order, they don't quite match
> 
> Screenshot 2023-05-02 at 9 28 01 AM

Updated

**Comment by ursaspace** [comment]
> Would it be helpful to add descriptions to these too? I think they're mostly self explanatory (except idk what Large Earth is), but might be nice to add extra context
> 
> Screenshot 2023-05-02 at 9 32 34 AM

There were descriptions for the primary analytics, but none for the sub-types for change detection. Checked both our internal and public facing documentation. I'd say its good enough for now, something we can add if requested.

File: README.md (hunk 2 above)
Comment (ursaspace) [reply README.md:50]: Sure, added a line that if a user is not logged in, tools requiring authentication will fail.

File: README.md (hunk 3 above)
Comment (ursaspace) [reply README.md]: Applied!

File: README.md (hunk 1 above)
Comment (ursaspace) [reply README.md]: Applied!

**Comment by ursaspace** [comment]
> When I empty either the start or end date, all the fields error, but that doesn't happen with any of the other fields, so that feels a little weird to me. I think it's fair to mark the field I messed with as invalid, but not the other ones (the field I actually touched it the only one that isn't showing errored)
> 
> Screenshot 2023-05-02 at 9 27 18 AM
> 
> here's the error message I get when i hover Screenshot 2023-05-02 at 9 38 29 AM

Ok, added some handling to check for NoneType. In this case, there doesn't need to be error messaging thrown since its a required field and arcpy parameters have their own validators for that. You can't run the tool until all required fields are provided. It will show a red asterisks until its filled out.

**Comment by samanthaweeks** [comment]
> > I know the start and end date have to be 2 days apart, but if I changed the start date to be later than the end date (rather than changing the end date) should we display the error on the start field since that's the one I edited? It can be the same exact message, because it's the same error, but I think we should error on whichever field the user actually touched
> 
> Two things:
> 
> 1. I think this is ok since (in this example) the user specifically wants that start date. The end date here is what is now invalid, not the start the date.
> 2. There isn't a good way to know which parameter actually changed. Whenever any one of the parameters change, those validators are ran. So in the validators, you can't know which value actually changed, just what the latest value from that parameter is, and you can't access previous values. I think thats just a limitation of these arcpy parameters. So in this example, we won't know whether the start date or the end date was what has changed.

Gotcha, that makes sense. I guess that's how we handle it in our UI too - we error on the end date regardless of what's changed. We _could_ error on both fields, but that seems excessive so yeah, i'll say this is resolved.

✅

**Pull request merged**